src/ripple/app/main/Application.cpp

```diff
@@ -1289,156 +1299,198 @@ ApplicationImp::getLastFullLedger()
-                             {
-                                 // VFALCO TODO This is the only place that
-                                 //             constructor is used, try to remove it
-                                 STLedgerEntry sle (*stp.object, uIndex);
-                                 bool ok = loadLedger->addSLE (sle);
-                                 if (!ok)
-                                 {
-                                     JLOG(m_journal.warn())
-                                        << "Couldn't add serialized ledger: "
-                                        << uIndex;
-                                 }
-                             }
-                             else
-                             {
-                                 JLOG(m_journal.warn())
-                                    << "Invalid entry in ledger";
-                             }
-                         }
-
-                         loadLedger->stateMap().flushDirty
-                             (hotACCOUNT_NODE, loadLedger->info().seq);
-                         loadLedger->setAccepted (closeTime,
-                             closeTimeResolution, ! closeTimeEstimated,
-                                *config_);
-                     }
-                 }
+                closeTimeResolution = std::chrono::seconds{
+                    ledger.get()["close_time_resolution"].asUInt()};
             }
+            if (ledger.get().isMember ("close_time_estimated"))
+            {
+                closeTimeEstimated =
+                    ledger.get()["close_time_estimated"].asBool();
+            }
+            if (ledger.get().isMember ("total_coins"))
+            {
+                totalDrops =
+                    beast::lexicalCastThrow<std::uint64_t>
+                        (ledger.get()["total_coins"].asString());
+            }
+
+            ledger = ledger.get()["accountState"];
         }
-        else if (ledgerID.empty () || (ledgerID == "latest"))
+
+        if (!ledger.get().isArray ())
         {
-            loadLedger = getLastFullLedger ();
+            JLOG(m_journal.fatal())
+               << "State nodes must be an array";
+            return nullptr;
+        }
+
+        auto loadLedger = std::make_shared<Ledger> (
+            seq, closeTime, *config_, family());
+        loadLedger->setTotalDrops(totalDrops);
+
+        for (Json::UInt index = 0; index < ledger.get().size(); ++index)
+        {
+            Json::Value& entry = ledger.get()[index];
+
+            if (!entry.isObject())
+            {
+                JLOG(m_journal.fatal())
+                    << "Invalid entry in ledger";
+                return nullptr;
+            }
+
+            uint256 uIndex;
+
+            if (!uIndex.SetHex (entry[jss::index].asString()))
+            {
+                JLOG(m_journal.fatal())
+                    << "Invalid entry in ledger";
+                return nullptr;
+            }
```

Comment:
Good extra checks in this loop. And big improvement to abort when a bad entry is found rather than the old continuation behavior, checking only at the end.